In [1]:
#Import Libraries that we may need for the project
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
#Import datasets and look at the data
critic_reviews=pd.read_csv("IMDb Movies/rotten_tomatoes_critic_reviews.csv")
critic_reviews.head()

rotten_tomatoes_link      critic_name  top_critic           publisher_name  \
0            m/0814255  Andrew L. Urban       False           Urban Cinefile   
1            m/0814255    Louise Keller       False           Urban Cinefile   
2            m/0814255              NaN       False      FILMINK (Australia)   
3            m/0814255     Ben McEachen       False  Sunday Mail (Australia)   
4            m/0814255      Ethan Alter        True       Hollywood Reporter   

  review_type review_score review_date  \
0       Fresh          NaN  2010-02-06   
1       Fresh          NaN  2010-02-06   
2       Fresh          NaN  2010-02-09   
3       Fresh        3.5/5  2010-02-09   
4      Rotten          NaN  2010-02-10   

                                      review_content  
0  A fantasy adventure that fuses Greek mythology...  
1  Uma Thurman as Medusa, the gorgon with a coiff...  
2  With a top-notch cast and dazzling special eff...  
3  Whether audiences will get behind The Lightnin...  
4  What's really lacking in The Lightning Thief i...

In [3]:
critic_reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1130017 entries, 0 to 1130016
Data columns (total 8 columns):
 #   Column                Non-Null Count    Dtype 
---  ------                --------------    ----- 
 0   rotten_tomatoes_link  1130017 non-null  object
 1   critic_name           1111488 non-null  object
 2   top_critic            1130017 non-null  bool  
 3   publisher_name        1130017 non-null  object
 4   review_type           1130017 non-null  object
 5   review_score          824081 non-null   object
 6   review_date           1130017 non-null  object
 7   review_content        1064211 non-null  object
dtypes: bool(1), object(7)
memory usage: 61.4+ MB


Noted missing values for critic_name, review_score and review_content.  Need decision on how to treat missing values, however those may be related to the year range that we will not be using.  Need to merge 2 rotten tomatoe files and make new df with only data pertaining to the years we want

In [5]:
rt_movies=pd.read_csv("IMDb Movies/rotten_tomatoes_movies.csv")
rt_movies.head(2)

rotten_tomatoes_link                                        movie_title  \
0            m/0814255  Percy Jackson & the Olympians: The Lightning T...   
1            m/0878835                                        Please Give   

                                          movie_info  \
0  Always trouble-prone, the life of teenager Per...   
1  Kate (Catherine Keener) and her husband Alex (...   

                                   critics_consensus content_rating  \
0  Though it may seem like just another Harry Pot...             PG   
1  Nicole Holofcener's newest might seem slight i...              R   

                                              genres          directors  \
0  Action & Adventure, Comedy, Drama, Science Fic...     Chris Columbus   
1                                             Comedy  Nicole Holofcener   

                                      authors  \
0  Craig Titley, Chris Columbus, Rick Riordan   
1                           Nicole Holofcener   

                                              actors original_release_date  \
0  Logan Lerman, Brandon T. Jackson, Alexandra Da...            2010-02-12   
1  Catherine Keener, Amanda Peet, Oliver Platt, R...            2010-04-30   

   ...      production_company  tomatometer_status tomatometer_rating  \
0  ...        20th Century Fox              Rotten               49.0   
1  ...  Sony Pictures Classics     Certified-Fresh               87.0   

  tomatometer_count  audience_status  audience_rating audience_count  \
0             149.0          Spilled             53.0       254421.0   
1             142.0          Upright             64.0        11574.0   

   tomatometer_top_critics_count  tomatometer_fresh_critics_count  \
0                             43                               73   
1                             44                              123   

   tomatometer_rotten_critics_count  
0                                76  
1                                19  

[2 rows x 22 columns]

In [6]:
rt_movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17712 entries, 0 to 17711
Data columns (total 22 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   rotten_tomatoes_link              17712 non-null  object 
 1   movie_title                       17712 non-null  object 
 2   movie_info                        17391 non-null  object 
 3   critics_consensus                 9134 non-null   object 
 4   content_rating                    17712 non-null  object 
 5   genres                            17693 non-null  object 
 6   directors                         17518 non-null  object 
 7   authors                           16170 non-null  object 
 8   actors                            17360 non-null  object 
 9   original_release_date             16546 non-null  object 
 10  streaming_release_date            17328 non-null  object 
 11  runtime                           17398 non-null  float64
 12  prod

Missing values noted in many variables.  Same comment as above, merge first then look at which year they pertain to. In addition, actors column has more than one actor per field, this may have to be separated, for team to discuss treatment.  we may end up dropping this column if imdb file has actors column that is easier to work with

In [7]:
rt_merged=rt_movies.merge(critic_reviews, left_index=True,right_index=True, how="outer", suffixes=('', '_y'), validate="m:1")
rt_merged.drop(rt_merged.filter(regex='_y$').columns.tolist(),axis=1, inplace=True)
rt_merged.head(2)


rotten_tomatoes_link                                        movie_title  \
0            m/0814255  Percy Jackson & the Olympians: The Lightning T...   
1            m/0878835                                        Please Give   

                                          movie_info  \
0  Always trouble-prone, the life of teenager Per...   
1  Kate (Catherine Keener) and her husband Alex (...   

                                   critics_consensus content_rating  \
0  Though it may seem like just another Harry Pot...             PG   
1  Nicole Holofcener's newest might seem slight i...              R   

                                              genres          directors  \
0  Action & Adventure, Comedy, Drama, Science Fic...     Chris Columbus   
1                                             Comedy  Nicole Holofcener   

                                      authors  \
0  Craig Titley, Chris Columbus, Rick Riordan   
1                           Nicole Holofcener   

                                              actors original_release_date  \
0  Logan Lerman, Brandon T. Jackson, Alexandra Da...            2010-02-12   
1  Catherine Keener, Amanda Peet, Oliver Platt, R...            2010-04-30   

   ... tomatometer_top_critics_count  tomatometer_fresh_critics_count  \
0  ...                          43.0                             73.0   
1  ...                          44.0                            123.0   

  tomatometer_rotten_critics_count      critic_name  top_critic  \
0                             76.0  Andrew L. Urban       False   
1                             19.0    Louise Keller       False   

   publisher_name review_type  review_score  review_date  \
0  Urban Cinefile       Fresh           NaN   2010-02-06   
1  Urban Cinefile       Fresh           NaN   2010-02-06   

                                      review_content  
0  A fantasy adventure that fuses Greek mythology...  
1  Uma Thurman as Medusa, the gorgon with a coiff...  

[2 rows x 29 columns]

In [9]:
rt_merged['original_release_date']=pd.to_datetime(rt_merged.original_release_date, format='%Y-%m-%d')
rt_merged.insert(loc=10, column='Year_of_Release', value=rt_merged.original_release_date.dt.year)
rt_merged.insert(loc=11, column='Month_of_Release', value=rt_merged.original_release_date.dt.month)
rt_merged.insert(loc=12, column='Day_of_Release', value=rt_merged.original_release_date.dt.day)

In [10]:
rt_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1130017 entries, 0 to 1130016
Data columns (total 32 columns):
 #   Column                            Non-Null Count    Dtype         
---  ------                            --------------    -----         
 0   rotten_tomatoes_link              17712 non-null    object        
 1   movie_title                       17712 non-null    object        
 2   movie_info                        17391 non-null    object        
 3   critics_consensus                 9134 non-null     object        
 4   content_rating                    17712 non-null    object        
 5   genres                            17693 non-null    object        
 6   directors                         17518 non-null    object        
 7   authors                           16170 non-null    object        
 8   actors                            17360 non-null    object        
 9   original_release_date             16546 non-null    datetime64[ns]
 10  Year_of_Release   

In [12]:
# number of missing values for Rotten_merged
na_rt_col= rt_merged.isnull().sum()
na_rt_col[na_rt_col != 0]

rotten_tomatoes_link                1112305
movie_title                         1112305
movie_info                          1112626
critics_consensus                   1120883
content_rating                      1112305
genres                              1112324
directors                           1112499
authors                             1113847
actors                              1112657
original_release_date               1113471
Year_of_Release                     1113471
Month_of_Release                    1113471
Day_of_Release                      1113471
streaming_release_date              1112689
runtime                             1112619
production_company                  1112804
tomatometer_status                  1112349
tomatometer_rating                  1112349
tomatometer_count                   1112349
audience_status                     1112753
audience_rating                     1112601
audience_count                      1112602
tomatometer_top_critics_count   

<font color='blue'>Hasti: you can see most of columns are NA and only critic_name has above 111000 rows. We can remove critic_name and then look at the NA data and if all columns are NA remove those rows
</font> 

In [11]:
rt_selected_years = rt_merged[(rt_merged["Year_of_Release"] >= 2015) & (rt_merged["Year_of_Release"] <= 2019)]
rt_selected_years.head(2)

rotten_tomatoes_link  movie_title  \
308         m/1001_grams   1001 Grams   
500        m/100_streets  100 Streets   

                                            movie_info  \
308  Marie, a Norwegian scientist, attends a semina...   
500  Three different stories covering infidelity, a...   

                                     critics_consensus content_rating genres  \
308                                                NaN             NR  Drama   
500  100 Streets strands its talented cast - led by...             NR  Drama   

        directors      authors  \
308    Bent Hamer   Bent Hamer   
500  Jim O'Hanlon  Leon Butler   

                                                actors original_release_date  \
308  Ane Dahl Torp, Laurent Stocker, Stein Winge, H...            2015-05-08   
500  Idris Elba, Gemma Arterton, Charlie Creed-Mile...            2017-01-13   

     ...  tomatometer_top_critics_count  tomatometer_fresh_critics_count  \
308  ...                            7.0                             18.0   
500  ...                           10.0                             16.0   

     tomatometer_rotten_critics_count    critic_name  top_critic  \
308                               4.0  Susan Granger       False   
500                              23.0            NaN       False   

           publisher_name review_type  review_score  review_date  \
308  www.susangranger.com       Fresh           5/5   2006-04-05   
500            Total Film      Rotten           1/5   2008-03-14   

                                        review_content  
308                                                NaN  
500  A beggar's banquet of milky CGI beasties and c...  

[2 rows x 32 columns]

In [76]:
rt_selected_years.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3036 entries, 308 to 17708
Data columns (total 32 columns):
 #   Column                            Non-Null Count  Dtype         
---  ------                            --------------  -----         
 0   rotten_tomatoes_link              3036 non-null   object        
 1   movie_title                       3036 non-null   object        
 2   movie_info                        2986 non-null   object        
 3   critics_consensus                 1766 non-null   object        
 4   content_rating                    3036 non-null   object        
 5   genres                            3030 non-null   object        
 6   directors                         2974 non-null   object        
 7   authors                           2781 non-null   object        
 8   actors                            2950 non-null   object        
 9   original_release_date             3036 non-null   datetime64[ns]
 10  Year_of_Release                   3036 non-nu

<font color='blue'>Hasti: I don't think we should limited the years, as number of observations after selecting years is 3036. After merging rotten_tomatoes and IMDB we will have 15822 obs including NA which need to be considered
</font> 

In [13]:
imdb_names=pd.read_csv("IMDb Movies\IMDb names.csv")
imdb_names.head(2)

imdb_name_id           name               birth_name  height  \
0    nm0000001   Fred Astaire  Frederic Austerlitz Jr.   177.0   
1    nm0000002  Lauren Bacall        Betty Joan Perske   174.0   

                                                 bio  \
0  Fred Astaire was born in Omaha, Nebraska, to J...   
1  Lauren Bacall was born Betty Joan Perske on Se...   

                                       birth_details date_of_birth  \
0               May 10, 1899 in Omaha, Nebraska, USA    1899-05-10   
1  September 16, 1924 in The Bronx, New York City...    1924-09-16   

                            place_of_birth  \
0                     Omaha, Nebraska, USA   
1  The Bronx, New York City, New York, USA   

                                       death_details date_of_death  \
0  June 22, 1987 in Los Angeles, California, USA ...    1987-06-22   
1  August 12, 2014 in New York City, New York, US...    2014-08-12   

                   place_of_death reason_of_death  \
0  Los Angeles, California, USA         pneumonia   
1  New York City, New York, USA            stroke   

                                      spouses_string  spouses  divorces  \
0  Robyn Smith  (27 June 1980 - 22 June 1987) (hi...        2         0   
1  Jason Robards  (4 July 1961 - 10 September 196...        2         1   

   spouses_with_children  children  
0                      1         2  
1                      2         3

Not sure if any of these variables add any value to our data and whether we should bother with them
<br> <font color='blue'>Hasti: Agreed, but we can keep it for further analysis if needed </font> 

In [13]:
imdb_names.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 297705 entries, 0 to 297704
Data columns (total 17 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   imdb_name_id           297705 non-null  object 
 1   name                   297705 non-null  object 
 2   birth_name             297705 non-null  object 
 3   height                 44681 non-null   float64
 4   bio                    204698 non-null  object 
 5   birth_details          110612 non-null  object 
 6   date_of_birth          110612 non-null  object 
 7   place_of_birth         103992 non-null  object 
 8   death_details          39933 non-null   object 
 9   date_of_death          39933 non-null   object 
 10  place_of_death         37038 non-null   object 
 11  reason_of_death        22694 non-null   object 
 12  spouses_string         45352 non-null   object 
 13  spouses                297705 non-null  int64  
 14  divorces               297705 non-nu

In [14]:
imdb_ratings=pd.read_csv("IMDb Movies\IMDb ratings.csv")
imdb_ratings.head()

imdb_title_id  weighted_average_vote  total_votes  mean_vote  median_vote  \
0     tt0000009                    5.9          154        5.9          6.0   
1     tt0000574                    6.1          589        6.3          6.0   
2     tt0001892                    5.8          188        6.0          6.0   
3     tt0002101                    5.2          446        5.3          5.0   
4     tt0002130                    7.0         2237        6.9          7.0   

   votes_10  votes_9  votes_8  votes_7  votes_6  ...  females_30age_avg_vote  \
0        12        4       10       43       28  ...                     5.7   
1        57       18       58      137      139  ...                     6.2   
2         6        6       17       44       52  ...                     5.8   
3        15        8       16       62       98  ...                     5.5   
4       210      225      436      641      344  ...                     7.3   

   females_30age_votes  females_45age_avg_vote  females_45age_votes  \
0                 13.0                     4.5                  4.0   
1                 23.0                     6.6                 14.0   
2                  4.0                     6.8                  7.0   
3                 14.0                     6.1                 21.0   
4                 82.0                     7.4                 77.0   

   top1000_voters_rating  top1000_voters_votes  us_voters_rating  \
0                    5.7                  34.0               6.4   
1                    6.4                  66.0               6.0   
2                    5.4                  32.0               6.2   
3                    4.9                  57.0               5.5   
4                    6.9                 139.0               7.0   

   us_voters_votes  non_us_voters_rating  non_us_voters_votes  
0             51.0                   6.0                 70.0  
1             96.0                   6.2                331.0  
2             31.0                   5.9                123.0  
3            207.0                   4.7                105.0  
4            488.0                   7.0               1166.0  

[5 rows x 49 columns]

In [15]:
imdb_ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85855 entries, 0 to 85854
Data columns (total 49 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   imdb_title_id              85855 non-null  object 
 1   weighted_average_vote      85855 non-null  float64
 2   total_votes                85855 non-null  int64  
 3   mean_vote                  85855 non-null  float64
 4   median_vote                85855 non-null  float64
 5   votes_10                   85855 non-null  int64  
 6   votes_9                    85855 non-null  int64  
 7   votes_8                    85855 non-null  int64  
 8   votes_7                    85855 non-null  int64  
 9   votes_6                    85855 non-null  int64  
 10  votes_5                    85855 non-null  int64  
 11  votes_4                    85855 non-null  int64  
 12  votes_3                    85855 non-null  int64  
 13  votes_2                    85855 non-null  int

same comments as above about the missing values. <br> <font color='blue'>Hasti: Here also some of the columns are mostly missing and we can drop these columns, next code </font> 

In [17]:
#number of missing values for IMDB-Rating
na_rating_col= imdb_ratings.isnull().sum()
na_rating_col[na_rating_col != 0]

allgenders_0age_avg_vote     52496
allgenders_0age_votes        52496
allgenders_18age_avg_vote      706
allgenders_18age_votes         706
allgenders_30age_avg_vote       10
allgenders_30age_votes          10
allgenders_45age_avg_vote       80
allgenders_45age_votes          80
males_allages_avg_vote           1
males_allages_votes              1
males_0age_avg_vote          58444
males_0age_votes             58444
males_18age_avg_vote          1465
males_18age_votes             1465
males_30age_avg_vote            12
males_30age_votes               12
males_45age_avg_vote           101
males_45age_votes              101
females_allages_avg_vote        81
females_allages_votes           81
females_0age_avg_vote        63738
females_0age_votes           63738
females_18age_avg_vote        6521
females_18age_votes           6521
females_30age_avg_vote         944
females_30age_votes            944
females_45age_avg_vote        2798
females_45age_votes           2798
top1000_voters_ratin

In [18]:
imdb_movies=pd.read_csv("IMDb Movies\IMDb movies.csv")
imdb_movies.head(2)

C:\Users\Hasty\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3071: DtypeWarning: Columns (3) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


imdb_title_id                        title               original_title  \
0     tt0000009                   Miss Jerry                   Miss Jerry   
1     tt0000574  The Story of the Kelly Gang  The Story of the Kelly Gang   

   year date_published                    genre  duration    country language  \
0  1894     1894-10-09                  Romance        45        USA     None   
1  1906     1906-12-26  Biography, Crime, Drama        70  Australia     None   

          director  ...                                             actors  \
0  Alexander Black  ...  Blanche Bayliss, William Courtenay, Chauncey D...   
1     Charles Tait  ...  Elizabeth Tait, John Tait, Norman Campbell, Be...   

                                         description avg_vote votes  budget  \
0  The adventures of a female reporter in the 1890s.      5.9   154     NaN   
1  True story of notorious Australian outlaw Ned ...      6.1   589  $ 2250   

   usa_gross_income worlwide_gross_income metascore reviews_from_users  \
0               NaN                   NaN       NaN                1.0   
1               NaN                   NaN       NaN                7.0   

   reviews_from_critics  
0                   2.0  
1                   7.0  

[2 rows x 22 columns]

In [19]:
# column year some of the years saved as string so I changed the type in next step
imdb_movies["year"].unique()

array([1894, 1906, 1911, 1912, 1919, 1913, 1914, 1915, 1916, 1917, 1918,
       1920, 1921, 1924, 1922, 1923, 1925, 1926, 1935, 1927, 1928, 1983,
       1929, 1930, 1932, 1931, 1937, 1938, 1933, 1934, 1936, 1940, 1939,
       1942, 1943, 1941, 1948, 1944, 2001, 1946, 1945, 1947, 1973, 1949,
       1950, 1952, 1951, 1962, 1953, 1954, 1955, 1961, 1956, 1958, 1957,
       1959, 1960, 1963, 1965, 1971, 1964, 1966, 1968, 1967, 1969, 1976,
       1970, 1979, 1972, 1981, 1978, 2000, 1989, 1975, 1974, 1986, 1990,
       2018, 1977, 1982, 1980, 1993, 1984, 1985, 1988, 1987, 2005, 1991,
       2002, 1994, 1992, 1995, 2017, 1997, 1996, 2006, 1999, 1998, 2007,
       2008, 2003, 2004, 2010, 2009, 2011, 2013, 2012, 2016, 2015, 2014,
       2019, 2020, '2012', '2015', '2009', '2013', '2018', '2014', '2017',
       '2011', '2016', '1981', '1975', '2010', '1984', '2007', '2006',
       '2001', '2004', '1979', '2019', '1967', '1978', '2003', '2005',
       '1969', '1990', '1983', '2002', '1996', '2008'

In [20]:
def convert_str_to_int(column):
    if pd.isnull(column) == True:
        return column
    if type(column) != str:
        return column
    else:
        string = column.split(' ')[-1]
        string = string.replace("'", '')
        #string = string.replace(',', '')
        return int(string)
    
imdb_movies["year"] = imdb_movies["year"].apply(convert_str_to_int)
imdb_movies["year"].unique()

array([1894, 1906, 1911, 1912, 1919, 1913, 1914, 1915, 1916, 1917, 1918,
       1920, 1921, 1924, 1922, 1923, 1925, 1926, 1935, 1927, 1928, 1983,
       1929, 1930, 1932, 1931, 1937, 1938, 1933, 1934, 1936, 1940, 1939,
       1942, 1943, 1941, 1948, 1944, 2001, 1946, 1945, 1947, 1973, 1949,
       1950, 1952, 1951, 1962, 1953, 1954, 1955, 1961, 1956, 1958, 1957,
       1959, 1960, 1963, 1965, 1971, 1964, 1966, 1968, 1967, 1969, 1976,
       1970, 1979, 1972, 1981, 1978, 2000, 1989, 1975, 1974, 1986, 1990,
       2018, 1977, 1982, 1980, 1993, 1984, 1985, 1988, 1987, 2005, 1991,
       2002, 1994, 1992, 1995, 2017, 1997, 1996, 2006, 1999, 1998, 2007,
       2008, 2003, 2004, 2010, 2009, 2011, 2013, 2012, 2016, 2015, 2014,
       2019, 2020], dtype=int64)

In [21]:
imdb_movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85855 entries, 0 to 85854
Data columns (total 22 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   imdb_title_id          85855 non-null  object 
 1   title                  85855 non-null  object 
 2   original_title         85855 non-null  object 
 3   year                   85855 non-null  int64  
 4   date_published         85855 non-null  object 
 5   genre                  85855 non-null  object 
 6   duration               85855 non-null  int64  
 7   country                85791 non-null  object 
 8   language               85022 non-null  object 
 9   director               85768 non-null  object 
 10  writer                 84283 non-null  object 
 11  production_company     81400 non-null  object 
 12  actors                 85786 non-null  object 
 13  description            83740 non-null  object 
 14  avg_vote               85855 non-null  float64
 15  vo

same comments about missing values.  here we can see that the actors column again contains multiple actors per field, need to clean up <br> <font color='blue'>Hasti: I think if we want to investigate on each actor, then we need to clean up this column, otherwise we can ignore it </font> 

In [22]:
# duplicated rows based on ACTORS column
imdb_movies[imdb_movies['actors'].duplicated()]

imdb_title_id                                              title  \
249       tt0012314  Das indische Grabmal erster Teil - Die Sendung...   
9212      tt0050718                          Il risveglio della mummia   
10607     tt0054663                            Bara roligt i Bullerbyn   
11105     tt0056060                             Heaven and Earth Magic   
13783     tt0063736                           Un film comme les autres   
...             ...                                                ...   
85328     tt9266104                               Zoku Owarimonogatari   
85489     tt9465532                                               Ogre   
85511     tt9496406                                       Feng yu zhou   
85537     tt9537054                                             Bijoya   
85769     tt9811598                                              Chaos   

                                          original_title  year date_published  \
249    Das indische Grabmal erster Teil - Die Sendung...  1921           1922   
9212                                     La momia azteca  1957     1957-11-13   
10607                            Bara roligt i Bullerbyn  1961     1961-12-16   
11105                             Heaven and Earth Magic  1962           1962   
13783                           Un film comme les autres  1968     1968-12-29   
...                                                  ...   ...            ...   
85328                               Zoku Owarimonogatari  2018     2018-11-10   
85489                                               Ogre  2019     2019-05-14   
85511                                       Feng yu zhou  2018     2018-08-03   
85537                                             Bijoya  2019     2019-01-04   
85769                                              Chaos  2019     2019-02-16   

                    genre  duration                            country  \
249    Adventure, Fantasy       120                            Germany   
9212               Horror        80                             Mexico   
10607   Adventure, Family        75                             Sweden   
11105           Animation        66                                USA   
13783               Drama       120                             France   
...                   ...       ...                                ...   
85328  Animation, Fantasy       148                              Japan   
85489           Animation        70                                USA   
85511           Animation       105                              China   
85537               Drama       145                              India   
85769      Drama, Fantasy       126  Russia, Ukraine, Austria, Tunisia   

               language         director  ...  \
249              German          Joe May  ...   
9212   Spanish, Nahuatl  Rafael Portillo  ...   
10607           Swedish     Olle Hellbom  ...   
11105           English      Harry Smith  ...   
13783            French  Jean-Luc Godard  ...   
...                 ...              ...  ...   
85328          Japanese   Akiyuki Shinbo  ...   
85489           English     James Snider  ...   
85511          Mandarin          Kuo Liu  ...   
85537           Bengali  Kaushik Ganguly  ...   
85769  English, Russian        Sergey A.  ...   

                                                  actors  \
249    Olaf Fønss, Mia May, Conrad Veidt, Erna Morena...   
9212   Ramón Gay, Rosita Arenas, Crox Alvarado, Luis ...   
10607  Kaj Andersson, Tove Hellbom, Jan Erik Husbom, ...   
11105                                                NaN   
13783                                                NaN   
...                                                  ...   
85328                                                NaN   
85489     Thomas Freeley, Maria Petrano, Jacob Whiteshed   
85511                                                NaN   
85537  Jaya Ahsan, Abir Chatterjee, Kaushik Ganguly, ...   
85769 

In [23]:
imdb_principals=pd.read_csv("IMDb Movies\IMDb title_principals.csv")
imdb_principals.head()

imdb_title_id  ordering imdb_name_id  category  job  \
0     tt0000009         1    nm0063086   actress  NaN   
1     tt0000009         2    nm0183823     actor  NaN   
2     tt0000009         3    nm1309758     actor  NaN   
3     tt0000009         4    nm0085156  director  NaN   
4     tt0000574         1    nm0846887   actress  NaN   

                                          characters  
0           ["Miss Geraldine Holbrook (Miss Jerry)"]  
1                                   ["Mr. Hamilton"]  
2  ["Chauncey Depew - the Director of the New Yor...  
3                                                NaN  
4                                     ["Kate Kelly"]

In [24]:
imdb_principals.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 835513 entries, 0 to 835512
Data columns (total 6 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   imdb_title_id  835513 non-null  object
 1   ordering       835513 non-null  int64 
 2   imdb_name_id   835513 non-null  object
 3   category       835513 non-null  object
 4   job            212731 non-null  object
 5   characters     340836 non-null  object
dtypes: int64(1), object(5)
memory usage: 38.2+ MB


I am unsure about the purpose of this file, I think Hasti was looking at it, does it join the movies to the actors across the common keys? <br> <font color='blue'>Hasti: Yes if we want to merge IMDB_names nd rating or movies, need this file, I add only imbd_name_id to the imdb_movies, so the rest we can ignore </font> 

In [25]:
# these mergings validate duplicate rows as m:1
imdb_movies_principal1=imdb_movies.merge(imdb_principals[['imdb_title_id', 'imdb_name_id']], left_index=True,right_index=True, how="outer", suffixes=('', '_y'), validate="m:1")
imdb_movies_principal1.drop(imdb_movies_principal1.filter(regex='_y$').columns.tolist(),axis=1, inplace=True)
imdb_movies_principal1

imdb_title_id                        title  \
0          tt0000009                   Miss Jerry   
1          tt0000574  The Story of the Kelly Gang   
2          tt0001892               Den sorte drøm   
3          tt0002101                    Cleopatra   
4          tt0002130                    L'Inferno   
...              ...                          ...   
835508           NaN                          NaN   
835509           NaN                          NaN   
835510           NaN                          NaN   
835511           NaN                          NaN   
835512           NaN                          NaN   

                     original_title    year date_published  \
0                        Miss Jerry  1894.0     1894-10-09   
1       The Story of the Kelly Gang  1906.0     1906-12-26   
2                    Den sorte drøm  1911.0     1911-08-19   
3                         Cleopatra  1912.0     1912-11-13   
4                         L'Inferno  1911.0     1911-03-06   
...                             ...     ...            ...   
835508                          NaN     NaN            NaN   
835509                          NaN     NaN            NaN   
835510                          NaN     NaN            NaN   
835511                          NaN     NaN            NaN   
835512                          NaN     NaN            NaN   

                            genre  duration           country language  \
0                         Romance      45.0               USA     None   
1         Biography, Crime, Drama      70.0         Australia     None   
2                           Drama      53.0  Germany, Denmark      NaN   
3                  Drama, History     100.0               USA  English   
4       Adventure, Drama, Fantasy      68.0             Italy  Italian   
...                           ...       ...               ...      ...   
835508                        NaN       NaN               NaN      NaN   
835509                        NaN       NaN               NaN      NaN   
835510                        NaN       NaN               NaN      NaN   
835511                        NaN       NaN               NaN      NaN   
835512                        NaN       NaN               NaN      NaN   

                                   director  ...  \
0                           Alexander Black  ...   
1                              Charles Tait  ...   
2                                 Urban Gad  ...   
3                        Charles L. Gaskill  ...   
4       Francesco Bertolini, Adolfo Padovan  ...   
...                                     ...  ...   
835508                                  NaN  ...   
835509                                  NaN  ...   
835510                                  NaN  ...   
835511                                  NaN  ...   
835512                                  NaN  ...   

                                              description avg_vote   votes  \
0       The adventures of a female reporter in the 1890s.      5.9   154.0   
1       True story of notorious Australian outlaw Ned ...      6.1   589.0   
2       Two men of high rank are both wooing the beaut...      5.8   188.0   
3       The fabled queen of Egypt's affair with Roman ...      5.2   446.0   
4       Loosely adapted from Dante's Divine Comedy and...      7.0  2237.0   
...                                                   ...      ...     ...   
835508                                                NaN      NaN     NaN   
835509                                                NaN      NaN     NaN   
835510                                                NaN      NaN     NaN   
835511                                                NaN      NaN     NaN   
835512                                                NaN      NaN     NaN   

         budget  usa_gross_income  worlwide_gross_income metascore  \
0           NaN               NaN                    NaN       NaN   
1        $ 2250               NaN 

<br> <font color='blue'>Hasti: first I did outer merge but you can see most of rows are all missed, then I tried inner and left merging and results are same, next step </font> 

In [26]:
imdb_movies_principal=imdb_movies.merge(imdb_principals[['imdb_title_id', 'imdb_name_id']], left_index=True,right_index=True, how="left", suffixes=('', '_y'), validate="m:1")
imdb_movies_principal.drop(imdb_movies_principal.filter(regex='_y$').columns.tolist(),axis=1, inplace=True)
imdb_movies_principal

imdb_title_id                           title  \
0         tt0000009                      Miss Jerry   
1         tt0000574     The Story of the Kelly Gang   
2         tt0001892                  Den sorte drøm   
3         tt0002101                       Cleopatra   
4         tt0002130                       L'Inferno   
...             ...                             ...   
85850     tt9908390                         Le lion   
85851     tt9911196  De Beentjes van Sint-Hildegard   
85852     tt9911774       Padmavyuhathile Abhimanyu   
85853     tt9914286               Sokagin Çocuklari   
85854     tt9914942      La vida sense la Sara Amat   

                       original_title  year date_published  \
0                          Miss Jerry  1894     1894-10-09   
1         The Story of the Kelly Gang  1906     1906-12-26   
2                      Den sorte drøm  1911     1911-08-19   
3                           Cleopatra  1912     1912-11-13   
4                           L'Inferno  1911     1911-03-06   
...                               ...   ...            ...   
85850                         Le lion  2020     2020-01-29   
85851  De Beentjes van Sint-Hildegard  2020     2020-02-13   
85852       Padmavyuhathile Abhimanyu  2019     2019-03-08   
85853               Sokagin Çocuklari  2019     2019-03-15   
85854      La vida sense la Sara Amat  2019     2020-02-05   

                           genre  duration           country       language  \
0                        Romance        45               USA           None   
1        Biography, Crime, Drama        70         Australia           None   
2                          Drama        53  Germany, Denmark            NaN   
3                 Drama, History       100               USA        English   
4      Adventure, Drama, Fantasy        68             Italy        Italian   
...                          ...       ...               ...            ...   
85850                     Comedy        95   France, Belgium         French   
85851              Comedy, Drama       103       Netherlands  German, Dutch   
85852                      Drama       130             India      Malayalam   
85853              Drama, Family        98            Turkey        Turkish   
85854                      Drama        74             Spain        Catalan   

                                  director  ...  \
0                          Alexander Black  ...   
1                             Charles Tait  ...   
2                                Urban Gad  ...   
3                       Charles L. Gaskill  ...   
4      Francesco Bertolini, Adolfo Padovan  ...   
...                                    ...  ...   
85850               Ludovic Colbeau-Justin  ...   
85851                      Johan Nijenhuis  ...   
85852                      Vineesh Aaradya  ...   
85853                    Ahmet Faik Akinci  ...   
85854                            Laura Jou  ...   

                                             description avg_vote votes  \
0      The adventures of a female reporter in the 1890s.      5.9   154   
1      True story of notorious Australian outlaw Ned ...      6.1   589   
2      Two men of high rank are both wooing the beaut...      5.8   188   
3      The fabled queen of Egypt's affair with Roman ...      5.2   446   
4      Loosely adapted from Dante's Divine Comedy and...      7.0  2237   
...                                                  ...      ...   ...   
85850  A psychiatric hospital patient pretends to be ...      5.3   398   
85851  A middle-aged veterinary surgeon believes his ...      7.7   724   
85852                                                NaN      7.9   265   
85853                                                NaN      6.4   194   
85854  Pep, a 13-year-old boy, is in love with a girl...      6.7   102   

        budget  usa_gross_income  worlwide_gross_income metascore  \
0          NaN               NaN                    NaN       NaN   

In [27]:
# merging imdb movies and ratings including imdb_names_id column
imdb_merged=imdb_movies.merge(imdb_ratings, left_index=True,right_index=True, how="outer", suffixes=('', '_y'), validate="m:1")
imdb_merged.drop(imdb_merged.filter(regex='_y$').columns.tolist(),axis=1, inplace=True)
imdb_merged

imdb_title_id                           title  \
0         tt0000009                      Miss Jerry   
1         tt0000574     The Story of the Kelly Gang   
2         tt0001892                  Den sorte drøm   
3         tt0002101                       Cleopatra   
4         tt0002130                       L'Inferno   
...             ...                             ...   
85850     tt9908390                         Le lion   
85851     tt9911196  De Beentjes van Sint-Hildegard   
85852     tt9911774       Padmavyuhathile Abhimanyu   
85853     tt9914286               Sokagin Çocuklari   
85854     tt9914942      La vida sense la Sara Amat   

                       original_title  year date_published  \
0                          Miss Jerry  1894     1894-10-09   
1         The Story of the Kelly Gang  1906     1906-12-26   
2                      Den sorte drøm  1911     1911-08-19   
3                           Cleopatra  1912     1912-11-13   
4                           L'Inferno  1911     1911-03-06   
...                               ...   ...            ...   
85850                         Le lion  2020     2020-01-29   
85851  De Beentjes van Sint-Hildegard  2020     2020-02-13   
85852       Padmavyuhathile Abhimanyu  2019     2019-03-08   
85853               Sokagin Çocuklari  2019     2019-03-15   
85854      La vida sense la Sara Amat  2019     2020-02-05   

                           genre  duration           country       language  \
0                        Romance        45               USA           None   
1        Biography, Crime, Drama        70         Australia           None   
2                          Drama        53  Germany, Denmark            NaN   
3                 Drama, History       100               USA        English   
4      Adventure, Drama, Fantasy        68             Italy        Italian   
...                          ...       ...               ...            ...   
85850                     Comedy        95   France, Belgium         French   
85851              Comedy, Drama       103       Netherlands  German, Dutch   
85852                      Drama       130             India      Malayalam   
85853              Drama, Family        98            Turkey        Turkish   
85854                      Drama        74             Spain        Catalan   

                                  director  ... females_30age_avg_vote  \
0                          Alexander Black  ...                    5.7   
1                             Charles Tait  ...                    6.2   
2                                Urban Gad  ...                    5.8   
3                       Charles L. Gaskill  ...                    5.5   
4      Francesco Bertolini, Adolfo Padovan  ...                    7.3   
...                                    ...  ...                    ...   
85850               Ludovic Colbeau-Justin  ...                    5.7   
85851                      Johan Nijenhuis  ...                    8.0   
85852                      Vineesh Aaradya  ...                    NaN   
85853                    Ahmet Faik Akinci  ...                    NaN   
85854                            Laura Jou  ...                    6.8   

      females_30age_votes females_45age_avg_vote females_45age_votes  \
0                    13.0                    4.5                 4.0   
1                    23.0                    6.6                14.0   
2                     4.0                    6.8                 7.0   
3                    14.0                    6.1                21.0   
4                    82.0                    7.4                77.0   
...                   ...                    ...                 ...   
85850                11.0                    5.0                 2.0   
85851                47.0                    7.3                30.0   
85852                 NaN                    NaN                 NaN   
85853                 NaN                 

In [28]:
#missing values for IMDB_merged
null_col= imdb_merged.isnull().sum()
null_col[null_col != 0]

country                         64
language                       833
director                        87
writer                        1572
production_company            4455
actors                          69
description                   2115
budget                       62145
usa_gross_income             70529
worlwide_gross_income        54839
metascore                    72550
reviews_from_users            7597
reviews_from_critics         11797
allgenders_0age_avg_vote     52496
allgenders_0age_votes        52496
allgenders_18age_avg_vote      706
allgenders_18age_votes         706
allgenders_30age_avg_vote       10
allgenders_30age_votes          10
allgenders_45age_avg_vote       80
allgenders_45age_votes          80
males_allages_avg_vote           1
males_allages_votes              1
males_0age_avg_vote          58444
males_0age_votes             58444
males_18age_avg_vote          1465
males_18age_votes             1465
males_30age_avg_vote            12
males_30age_votes   

<br> <font color='blue'>Hasti: Here also some of the columns are mostly missing and we can drop these columns <br> It depends which votes columns we want to use or consider </font> 

In [72]:
# there is no fully duplicated rows in the file
imdb_merged[imdb_merged.duplicated(subset=None, keep='first')]

Empty DataFrame
Columns: [imdb_title_id, title, original_title, year, date_published, genre, duration, country, language, director, writer, production_company, actors, description, avg_vote, votes, budget, usa_gross_income, worlwide_gross_income, metascore, reviews_from_users, reviews_from_critics, weighted_average_vote, total_votes, mean_vote, median_vote, votes_10, votes_9, votes_8, votes_7, votes_6, votes_5, votes_4, votes_3, votes_2, votes_1, allgenders_0age_avg_vote, allgenders_0age_votes, allgenders_18age_avg_vote, allgenders_18age_votes, allgenders_30age_avg_vote, allgenders_30age_votes, allgenders_45age_avg_vote, allgenders_45age_votes, males_allages_avg_vote, males_allages_votes, males_0age_avg_vote, males_0age_votes, males_18age_avg_vote, males_18age_votes, males_30age_avg_vote, males_30age_votes, males_45age_avg_vote, males_45age_votes, females_allages_avg_vote, females_allages_votes, females_0age_avg_vote, females_0age_votes, females_18age_avg_vote, females_18age_votes, females_30age_avg_vote, females_30age_votes, females_45age_avg_vote, females_45age_votes, top1000_voters_rating, top1000_voters_votes, us_voters_rating, us_voters_votes, non_us_voters_rating, non_us_voters_votes]
Index: []

[0 rows x 70 columns]

In [38]:
# merging IMDB and Rotten_Tomatoes by considering name of movies 
imdb_rt_merged=imdb_merged.merge(rt_merged, left_on=['original_title'],right_on=['movie_title'], how="inner", validate="m:m").drop('movie_title', axis=1)
imdb_rt_merged

imdb_title_id                                          title  \
0         tt0002101                                      Cleopatra   
1         tt0007801                                      Cleopatra   
2         tt0024991                                      Cleopatra   
3         tt0056937                                      Cleopatra   
4         tt0346765                                      Cleopatra   
...             ...                                            ...   
15817     tt9779804                                       15 Years   
15818     tt9860728  Falling Inn Love - Ristrutturazione con amore   
15819     tt9894470                                            VFW   
15820     tt9898858                                Coffee & Kareem   
15821     tt9904802                                    Enemy Lines   

         original_title  year date_published                      genre  \
0             Cleopatra  1912     1912-11-13             Drama, History   
1             Cleopatra  1917     1917-10-14  Biography, Drama, History   
2             Cleopatra  1934     1934-10-05  Biography, Drama, History   
3             Cleopatra  1963     1964-01-30  Biography, Drama, History   
4             Cleopatra  2003     2003-08-14              Comedy, Drama   
...                 ...   ...            ...                        ...   
15817          15 Years  2019     2020-02-27                      Drama   
15818  Falling Inn Love  2019     2019-08-29            Comedy, Romance   
15819               VFW  2019     2020-02-14      Action, Crime, Horror   
15820   Coffee & Kareem  2020     2020-04-03             Action, Comedy   
15821       Enemy Lines  2020     2020-05-04                        War   

       duration               country                          language  \
0           100                   USA                           English   
1           125                   USA                           English   
2           100                   USA                           English   
3           192  Switzerland, UK, USA                           English   
4           104      Spain, Argentina                           Spanish   
...         ...                   ...                               ...   
15817        89                Israel                            Hebrew   
15818        98                   USA                           English   
15819        92                   USA                           English   
15820        88                   USA                           English   
15821        92                    UK  English, Polish, Russian, German   

                   director  ... tomatometer_top_critics_count  \
0        Charles L. Gaskill  ...                          10.0   
1         J. Gordon Edwards  ...                          10.0   
2          Cecil B. DeMille  ...                          10.0   
3      Joseph L. Mankiewicz  ...                          10.0   
4          Eduardo Mignogna  ...                          10.0   
...                     ...  ...                           ...   
15817          Yuval Hadadi  ...                           1.0   
15818          Roger Kumble  ...                           3.0   
15819             Joe Begos  ...                           3.0   
15820         Michael Dowse  ...                          17.0   
15821          Anders Banke  ...                           1.0   

      tomatometer_fresh_critics_count tomatometer_rotten_critics_count  \
0                                24.0                             15.0   
1                                24.0                             15.0   
2                                24.0                             15.0   
3                                24.0                             15.0   
4                                24.0                             15.0   
...                               ...                              ...   
15817                             4.0               

In [35]:
pd.set_option('display.max_rows', None)
na_col_t= imdb_rt_merged.isnull().sum()
na_col_t[na_col_t != 0]

country                         6
language                      103
director                        6
writer                         55
production_company_x          217
actors_x                        4
description                    35
budget                       6934
usa_gross_income             6561
worlwide_gross_income        5306
metascore                    6289
reviews_from_users            126
reviews_from_critics          306
allgenders_0age_avg_vote     4523
allgenders_0age_votes        4523
allgenders_18age_avg_vote      20
allgenders_18age_votes         20
allgenders_45age_avg_vote       5
allgenders_45age_votes          5
males_allages_avg_vote          1
males_allages_votes             1
males_0age_avg_vote          5703
males_0age_votes             5703
males_18age_avg_vote           56
males_18age_votes              56
males_30age_avg_vote            1
males_30age_votes               1
males_45age_avg_vote            5
males_45age_votes               5
females_0age_a

In [37]:
pd.reset_option('display.max_rows')

In [39]:
col_to_keep_rt= ['review_score', 'tomatometer_status', 'tomatometer_rating', 'tomatometer_count', 'audience_status', 
              'audience rating', 'audience_count', 'tomatometer_top_critics_count', 'tomatometer_fresh_critics_count',
              'tomatometer_rotten_critics_count']


In [40]:
#saved merged of two sources in imdn_rt_merged.csv
imdb_rt_merged.to_csv('imdb_rt_merged.csv', index=False)  

 <font color='blue'>Hasti: I didn't remove any column or rows.<br> Also the merged file included all years. It's easy to change if you want<br> Let me know if I should do something else  </font> 

In [43]:
import json
data_df = pd.read_json(json.dumps(imdb_rt_merged))
data_df

TypeError: Object of type DataFrame is not JSON serializable

In [41]:
import pickle
with open('imdb_rt_merged.pkl', 'wb') as handle:
    pickle.dump(imdb_rt_merged.csv, handle, protocol=pickle.HIGHEST_PROTOCOL)

pd.read_pickle('imdb_rt_merged.pkl')

AttributeError: 'DataFrame' object has no attribute 'csv'